server/user.go

```diff
@@ -689,11 +699,86 @@ type userCacheEntry struct {
 	topics int
 }
 
+// Preserved update entry kept while we read the unread counter from the DB.
+type bufferedUpdate struct {
+	val int
+	inc bool
+}
+
+// Unread counter read result.
+type ioResult struct {
+	uid types.Uid
+	val int
+	err error
+}
+
+// Represents pending push notification receipt.
+type pendingReceipt struct {
+	// Number of unread counters currently being read from the DB.
+	pendingIOs int
+	// The index is needed by update and is maintained by the heap.Interface methods.
+	index int
+	// Underlying receipt.
+	rcpt *push.Receipt
+}
+
+// Pending pushes organized as a priority queue (priority = number of pending IOs).
+// This way we can quickly discover.
+type pendingReceiptsQueue []*pendingReceipt
```

Comment:
Please consider `type pendingReceiptsQueue []pendingReceipt`

Reply:
I need to keep it `[]*pendingRecept` so I can update the `pendingIOs` field correctly across `receiptQueue` and `perUserPendingReceipts`.
(when an IO completes, the `pendingIOs` field gets decremented for all pending receipts stored in `perUserPendingReceipts`).

Reply:
OK